cmd/snap-seccomp/main_test.go

```diff
@@ -129,6 +134,205 @@ func nativeEndian() binary.ByteOrder {
+
+	struct sock_fprog prog = {
+		.len = num_read / sizeof(struct sock_filter),
+		.filter = (struct sock_filter *)bpf,
+	};
+
+	// Set NNP to allow loading seccomp policy into the kernel without
+	// root
+	if (prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0)) {
+		perror("prctl(PR_NO_NEW_PRIVS, 1, 0, 0, 0)");
+		exit(1);
+	}
+
+	if (prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, &prog)) {
+		perror("prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, ...) failed");
+		exit(1);
+	}
+	return 0;
+}
+
+int main(int argc, char *argv[])
+{
+	int rc = 0;
+	if (argc < 2) {
+		fprintf(stderr, "Usage: %s <bpf file> [prog ...]\n", argv[0]);
+		return 1;
+	}
+
+	rc = sc_apply_seccomp_bpf(argv[1]);
+	if (rc || argc == 2)
+		return rc;
+
+	execv(argv[2], (char *const *)&argv[2]);
+	perror("execv failed");
+	return 1;
+}
+`)
+
+var seccompSyscallRunnerContent = []byte(`
+#define _GNU_SOURCE
+#include<unistd.h>
+#include<sys/syscall.h>
+#include<stdlib.h>
+int main(int argc, char **argv) {
+   int l[7];
+   for (int i=0; i<7;i++)
+      l[i] = atoi(argv[i+1]);
+   syscall(l[0], l[1], l[2], l[3], l[4], l[5], l[6]);
+   syscall(SYS_exit, 0, 0, 0, 0, 0, 0);
+}
+`)
+
+func lastKmsg() string {
+	output, err := exec.Command("dmesg").CombinedOutput()
+	if err != nil {
+		return err.Error()
+	}
+	l := strings.Split(string(output), "\n")
+	return l[len(l)-2]
+}
+
+func (s *snapSeccompSuite) SetUpSuite(c *C) {
+	// FIXME: we currently use a fork of x/net/bpf because of:
+	//   https://github.com/golang/go/issues/20556
+	// switch to x/net/bpf once we can simulate seccomp bpf there
+	bpf.VmEndianness = nativeEndian()
+
+	// build seccomp-load helper
+	s.seccompBpfLoader = filepath.Join(c.MkDir(), "seccomp_bpf_loader")
+	err := ioutil.WriteFile(s.seccompBpfLoader+".c", seccompBpfLoaderContent, 0644)
+	c.Assert(err, IsNil)
+	cmd := exec.Command("gcc", "-Werror", "-Wall", s.seccompBpfLoader+".c", "-o", s.seccompBpfLoader)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err = cmd.Run()
+	c.Assert(err, IsNil)
+
+	// build syscall-runner helper
+	s.seccompSyscallRunner = filepath.Join(c.MkDir(), "seccomp_syscall_runner")
+	err = ioutil.WriteFile(s.seccompSyscallRunner+".c", seccompSyscallRunnerContent, 0644)
+	c.Assert(err, IsNil)
+	cmd = exec.Command("gcc", "-Werror", "-Wall", "-static", s.seccompSyscallRunner+".c", "-o", s.seccompSyscallRunner, "-Wl,-static", "-static-libgcc")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err = cmd.Run()
+	c.Assert(err, IsNil)
+
+}
+
+func (s *snapSeccompSuite) runBpfInKernel(c *C, seccompWhitelist, bpfInput string, expected int) {
+	// Common syscalls we need to allow for a minimal statically linked
+	// c program.
+	//
+	// If we compile a test program for each test we can get aways with
+	// a even smaller set of syscalls: execve,exit essentially. But it
+	// means a much longer test run (30s vs 2s). Commit d288d89 contains
+	// the code for this.
+	common := `
+execve
+uname
+brk
+arch_prctl
+readlink
+access
+sysinfo
+exit
+# i386
+set_thread_area
```

Comment:
Yeah, I was about to write that this list is a great way to learn how linux userspace works across architectures. I don't mind having it but I bet this will bite us the first time this runs on s360 and you have to fix the tests.